src/app/getting-started/services/provider.service.ts

```diff
@@ -23,16 +28,43 @@ export class ProviderService {
    * @param redirect URL to be redirected to after successful account linking
    */
   linkAll(redirect: string): void {
-    this.link(null, redirect);
+        let openShiftLinkingRedirectUrl = this.getLegacyLinkingUrl("openshift-v3",redirect);
+        // after linking github, proceed with linking openshift-v3,
+        // hence passing openshift linking url as a redirect.
+        this.linkGitHub(openShiftLinkingRedirectUrl);
   }
 
   /**
    * Link a GitHub account to the user account
    *
    * @param redirect URL to be redirected to after successful account linking
    */
-  linkGitHub(redirect: string): void {
-    this.link("github", redirect);
+  linkGitHub(next: string): void{
+    let githubLinkURL = this.linkUrl + "?for=https://github.com&redirect=" +  encodeURIComponent(next) ;
+    console.log("attempting to connect github "+ githubLinkURL);
+    this.http
+    .get(githubLinkURL)
+    .map(response => {
+      // TODO: what happens to this when the response is not a pure json
+      let redirectInfo = response.json() as Link;
+      this.redirectToAuth(redirectInfo.redirect_location);
+      console.log(redirectInfo);
+    })
+    .catch((error) => {
+      console.log("error while linking github "+ githubLinkURL);
+      return this.handleError(error);
+    }).subscribe();
+  }
+  
+
+  getLegacyLinkingUrl(provider: string, redirect: string): string{
+    let parsedToken = jwt_decode(this.auth.getToken());    
+    let url = this.loginUrl + "/session?"
+    + "clientSession=" + parsedToken.client_session
+    + "&sessionState=" + parsedToken.session_state
+    + "&redirect=" + redirect // brings us back to Getting Started.
+    + "&provider=" + provider;
+    return url;
```

Comment:
1. I'm still curious about the getLegacyLinkingUrl() name. Why is it a legacy URL?
2. It appears that the getLegacyLinkingUrl() and link() methods are nearly identical. Can the link method get its URL from getLegacyLinkingUrl() instead?
3. Are we still using the link() method (e.g., to link openshift)?

Reply:
link() ? There is no link() function method anymore. Could you elaborate please?
There are linkAll() links both OSO and GitHub
linkGitHub() uses new API to link GH
getLegacyLinkingUrl() to get a link to the legacy/depricated API which we will remove later.

They all are different.

Reply:
I was referring to the link() method in provider.service here: https://github.com/sbose78/fabric8-ui/blob/f84fad16eab2e2f0969163012e0873fa169af709/src/app/getting-started/services/provider.service.ts#L85

The code (below) to generate the url appears duplicated in getLegacyLinkingUrl() and link().

```
let parsedToken = jwt_decode(this.auth.getToken());    
    let url = this.loginUrl + "/session?"
    + "clientSession=" + parsedToken.client_session
    + "&sessionState=" + parsedToken.session_state
    + "&redirect=" + redirect // brings us back to Getting Started.
+ "&provider=" + provider;
```

Reply:
Oh, right. I missed the old link() :)
Yes, we should reuse getLegacyLinkUrl() in linkOpenShift() and remove link(). Good catch.

Reply:
Since this service will again undergo refactoring ( and there aren't too many tests ), I was avoiding touching the existing code :) Let me do it nevertheless. Thanks.

Reply:
Done.